autode/opt/optimisers/trm.py

```diff
@@ -0,0 +1,493 @@
+
+        rfo_h_eff = self._get_rfo_minimise_h_eff(self._coords)
+        rfo_step, rfo_step_size = self._get_step_and_cart_size_from_h_eff(
+            self._coords, rfo_h_eff
+        )
+
+        if rfo_step_size < self.alpha:
+            logger.info("Taking a pure RFO step")
+            step = rfo_step
+        else:
+            try:
+                qa_h_eff = self._get_trm_minimise_h_eff(self._coords)
+                qa_step, _ = self._get_step_and_cart_size_from_h_eff(
+                    self._coords, qa_h_eff
+                )
+                logger.info("Taking a TRM/QA step optimised to trust radius")
+                step = qa_step
+
+            except OptimiserStepError as exc:
+                logger.debug(f"TRM/QA step failed: {str(exc)}")
+                # if TRM/QA failed, take scaled RFO step
+                scaled_step = rfo_step * (self.alpha / rfo_step_size)
+
+                logger.info(
+                    "Taking a scaled RFO step approximately within "
+                    "trust radius"
+                )
+                step = scaled_step
+
+        self._coords.allow_unconverged_back_transform = False
+        self._coords = self._coords + step
+
+        step_size = np.linalg.norm(
+            self._coords.to("cart") - self._history.penultimate.to("cart")
+        )
+        logger.info(f"Size of step taken (in Cartesian) = {step_size:.3f} Å")
+        return None
+
+    @staticmethod
+    def _get_step_and_cart_size_from_h_eff(
+        old_coords: OptCoordinates, hess_eff: np.ndarray
+    ) -> Tuple[np.ndarray, float]:
+        """
+        Obtains the Cartesian step size given a step in the current
+        coordinate system (e.g., Delocalised Internal Coordinates)
+
+        Args:
+            old_coords (OptCoordinates): previous coordinates
+            hess_eff (np.ndarray): Effective (shifted) hessian
+
+        Returns:
+            (float): The step size in Cartesian
+        """
+        step = np.matmul(-np.linalg.inv(hess_eff), old_coords.g)
+        new_coords = old_coords + step
+        cart_delta = new_coords.to("cart") - old_coords.to("cart")
+        return step, float(np.linalg.norm(cart_delta))
+
+    @staticmethod
+    def _get_rfo_minimise_h_eff(coords) -> np.ndarray:
+        """
+        Using current Hessian and gradient, obtain the level-shifted
+        Hessian that would provide a minimising RFO step
+
+        Returns:
+            (np.ndarray): The level-shifted effective Hessian
+        """
+        h_n = coords.h.shape[0]
+
+        # form the augmented Hessian
+        aug_h = np.zeros(shape=(h_n + 1, h_n + 1))
+        aug_h[:h_n, :h_n] = coords.h
+        aug_h[-1, :h_n] = coords.g
+        aug_h[:h_n, -1] = coords.g
+
+        aug_h_lmda, aug_h_v = np.linalg.eigh(aug_h)
+
+        # RFO step uses the lowest non-zero eigenvalue
+        mode = np.where(np.abs(aug_h_lmda) > 1.0e-15)[0][0]
+        assert mode == 0
```

Comment:
@t-young31 yes I intentionally left it, because I think the problem is that if one or more eigenvalues are 0, then this kind of TRM or RFO step calculation with the lambda would not really work. You would need to project the gradient onto the Hessian eigenvectors and then remove the terms coming from 0 eigenvectors and then sum (f / b - lambda) and then convert it back to the step (like what is done in CRFO or PRFO). But this is too complicated, and would require much more code. I have checked other codes like pysisyphus and geomeTRIC and they all use this eigenvector method.

From my tests, even for Cartesian you don't get 0 eigenvalues anywhere, due to coupling between the translation/rotation I think. So I put that there to ensure that the optimiser stops if it encounters that scenario instead of quietly giving garbage results (because if the first eigenvalue is 0, then the step will go up in that eigenvalue with the lambda method). For internal coordinates its impossible for any eigenvalue to be 0 without constraints. What would you suggest with this?

Reply:
thanks for the detailed reply. In that case, would it be more clear to do something like:

```suggestion
        mode = 0
        if aug_h_lmda[mode] > -1E-15:
            raise OptimiserStepError("<why>")
```

Reply:
Yes that works, changed it